base/broadcast.jl

```diff
@@ -890,6 +890,30 @@ preprocess_args(dest, args::Tuple{}) = ()
     return dest
 end
 
+# Performance optimization: for BitVector outputs, we cache the result
+# in a 64-bit register before writing into memory (to bypass LSQ)
+@inline function copyto!(dest::BitVector, bc::Broadcasted{Nothing})
+    axes(dest) == axes(bc) || throwdm(axes(dest), axes(bc))
+    ischunkedbroadcast(dest, bc) && return chunkedcopyto!(dest, bc)
+    destc = dest.chunks
+    length(destc)<=0 && return dest
+    bcp = preprocess(dest, bc)
+    @inbounds for i = 0:length(destc)-2
+        z = UInt64(0)
+        for j=0:63
+           z |= (bcp[i*64 + j + 1]::Bool) << (j&63)
```

Comment:
Why the `& 63`?

Reply:
It avoids emitting code that checks if the shift value is between 0 and 63.

Reply:
Oh right, I see what you mean, it should probably be `&5` or something like that.

Reply:
I figured as such — in that case 63 is correct.

```llvm
julia> f(x, y) = x << y
f (generic function with 1 method)

julia> g(x, y) = x << (y & 63)
g (generic function with 1 method)

julia> @code_llvm debuginfo=:none f(UInt(1), 1)

define i64 @julia_f_15983(i64, i64) {
top:
  %2 = icmp slt i64 %1, 0
  %3 = icmp ugt i64 %1, 63
  %4 = shl i64 %0, %1
  %5 = select i1 %3, i64 0, i64 %4
  %6 = sub i64 0, %1
  %7 = icmp ugt i64 %6, 63
  %8 = lshr i64 %0, %6
  %9 = select i1 %7, i64 0, i64 %8
  %10 = select i1 %2, i64 %9, i64 %5
  ret i64 %10
}

julia> @code_llvm debuginfo=:none g(UInt(1), 1)

define i64 @julia_g_15989(i64, i64) {
top:
  %2 = and i64 %1, 63
  %3 = shl i64 %0, %2
  ret i64 %3
}
```

Reply:
As @maxbennedich recently pointed out, `f(a,b)=a<<b` and `g(a,b)= a<<(b&63)` compile to quite different code for integers on x86. If the second argument is larger or equal than the bitwidth, then julia wants to make the value zero, while the native instruction only considers the relevant bits (implements `g`). Hence, the `g` variant is faster on x86 if the compiler fails to prove that the second argument is between 0 and 63. 

What I really want here is`shl(a::UInt64,b::Int64)=Core.Intrinsics.llvmcall("%res =shl i64 %0, %1\n ret i64 %res", UInt64, Tuple{UInt64, Int64}, a, b)`, i.e. I want to express that shifting left by more than 63 bit is undefined behavior and llvm can emit whatever instruction does the job fastest on the target arch. I don't know what kind of idiom wins for arm or power.

You are definitely right that this is very smelly code.